File: src/main/java/ru/ioleksiv/telegram/bot/core/model/method/file/SendAnimationFile.java
package ru.ioleksiv.telegram.bot.core.model.method.file;

import org.springframework.core.io.FileSystemResource;
import ru.ioleksiv.telegram.bot.core.controller.network.Networker;
import ru.ioleksiv.telegram.bot.core.model.objects.std.keyboard.IKeyboard;

/**
 * @see <a href="https://core.telegram.org/bots/api#sendanimation>sendAnimation</a>
 */
public class SendAnimationFile extends UploadFile {
    private static final String METHOD = "sendAnimation";

    /**
     * animation	InputFile or String	Yes	Animation to send. Pass a file_id as String to send an
     * animation that exists on the Telegram servers (recommended), pass an HTTP URL as a String
     * for Telegram to get an animation from the Internet, or upload a new animation using
     * multipart/form-data.
     */
    private static final String ANIMATION_KEY = "animation";
    /**
     * duration	Integer	Optional	Duration of sent animation in seconds
     */
    private static final String DURATION_KEY = "duration";
    /**
     * width	Integer	Optional	Animation width
     */
    private static final String WIDTH_KEY = "width";
    /**
     * height	Integer	Optional	Animation height
     */
    private static final String HEIGHT_KEY = "height";
    /**
     * thumb	InputFile or String	Optional	Thumbnail of the file sent. The thumbnail should be
     * in JPEG format and less than 200 kB in size. A thumbnail‘s width and height should not exceed
     * 90. Ignored if the file is not uploaded using multipart/form-data. Thumbnails can’t be reused
     * and can be only uploaded as a new file, so you can pass “attach://<file_attach_name>” if the
     * thumbnail was uploaded using multipart/form-data under <file_attach_name>.
     */
    private static final String THUMB_KEY = "thumb";

    public SendAnimationFile(Networker networker) {
        super(METHOD, networker);
    }

    public SendAnimationFile setThumb(FileSystemResource thumb) {
        putFile(THUMB_KEY, thumb);
        return this;
    }

    public SendAnimationFile setAnimation(FileSystemResource animation) {
        putFile(ANIMATION_KEY, animation);
        return this;
    }

    public SendAnimationFile setDuration(Integer duration) {
        putInt(DURATION_KEY, duration);
        return this;
    }

    public SendAnimationFile setWidth(Integer width) {
        putInt(WIDTH_KEY, width);
        return this;
    }

    public SendAnimationFile setHeight(Integer height) {
        putInt(HEIGHT_KEY, height);
        return this;
    }

    public SendAnimationFile setReplyMarkup(IKeyboard replyMarkup) {
        putReplyMarkup(replyMarkup);
        return this;
    }

    public SendAnimationFile setDisableNotification(Boolean disableNotification) {
        putBool(DISABLE_NOTIFICATION_KEY, disableNotification);
        return this;
    }

    public SendAnimationFile setReplyToMessageId(Long replyToMessageId) {
        putLong(REPLY_TO_MESSAGE_ID, replyToMessageId);
        return this;
    }

    public SendAnimationFile setChatId(Long chatId) {
        putLong(CHAT_ID_KEY, chatId);
        return this;
    }

    public SendAnimationFile setParseMode(String parseMode) {
        putString(PARSE_MODE_KEY, parseMode);
        return this;
    }

    public SendAnimationFile setCaption(String caption) {
        putString(CAPTION_KEY, caption);
        return this;
    }

}

File: src/main/kotlin/io/github/oleksivio/telegram/bot/api/model/method/file/SendAnimationFile.kt
package io.github.oleksivio.telegram.bot.api.model.method.file

import com.fasterxml.jackson.annotation.JsonProperty
import io.github.oleksivio.telegram.bot.api.model.objects.std.keyboard.IKeyboard
import io.github.oleksivio.telegram.bot.core.model.ApiDict
import io.github.oleksivio.telegram.bot.core.model.method.ActionMap
import io.github.oleksivio.telegram.bot.core.model.method.file.UploadMediaFile
import org.springframework.core.io.FileSystemResource

/**
 * @see [sendAnimation](https://core.telegram.org/bots/api/#sendanimation)
 */
data class SendAnimationFile(
        /**
         * chat_id Integer Yes Unique identifier for the target chat
         */
        @JsonProperty(ApiDict.CHAT_ID_KEY)
        val chatId: Long,
        /**
         * animation InputFile or String Yes Animation to send. Pass a file_id as String to send an animation that exists on
         * the Telegram servers (recommended), pass an HTTP URL as a String for Telegram to get an animation from the
         * Internet, or upload a new animation using multipart/form-data.
         */
        val animation: FileSystemResource,
        /**
         * duration [Integer] OptionalDuration of sent animation in seconds
         */
        val duration: Int? = null,
        /**YY
         * width [Integer] OptionalAnimation width
         */
        val width: Int? = null,
        /**
         * height [Integer] OptionalAnimation height
         */
        val height: Int? = null,
        /**
         * thumb  [String] Optional Thumbnail of the file sent;
         * can be ignored if thumbnail generation for the file is supported server-side.
         * The thumbnail should be in JPEG format and less than 200 kB in size.
         * A thumbnail‘s width and height should not exceed 90.
         * Ignored if the file is not uploaded using multipart/form-data.
         * Thumbnails can’t be reused and can be only uploaded as a new file,
         * so you can pass “attach://<file_attach_name>” if the thumbnail was uploaded using multipart/form-data under <file_attach_name>. More info on Sending Files »
         */
        val thumb: FileSystemResource? = null,
        /**
         * caption [String] Optional Animation caption (may also be used when resending animation by file_id), 0-1024 characters
         */
        override val caption: String? = null,
        /**
         * parse_mode [String] Optional Send Markdown or HTML, if you want Telegram apps to show bold, italic,
         * fixed-width text or inline URLs in the media caption.
         */
        override val parseMode: String? = null,
        /**
         * reply_markup InlineKeyboardMarkup or ReplyKeyboardMarkup or ReplyKeyboardRemove or ForceReply Optional
         * Additional interface options. A JSON-serialized object for an inline keyboard, custom reply keyboard,
         * instructions to remove reply keyboard or to force a reply from the user.
         */
        override val replyMarkup: IKeyboard? = null,
        /**
         * disable_notification [Boolean] OptionalSends the message silently.
         * Users will receive a notification with no sound.
         */
        override val disableNotification: Boolean? = null,
        /**
         * reply_to_message_id [Integer] OptionalIf the message is a reply, ID of the original message
         */
        override val replyToMessageId: Long? = null
) : UploadMediaFile() {

    override fun ActionMap.fill() {
        thumb?.let { putFile(ApiDict.THUMB_KEY, it) }
        putFile(ApiDict.ANIMATION_KEY, animation)
        duration?.let { putInt(ApiDict.DURATION_KEY, it) }
        width?.let { putInt(ApiDict.WIDTH_KEY, it) }
        height?.let { putInt(ApiDict.HEIGHT_KEY, it) }
        replyMarkup?.let { putObject(ApiDict.REPLY_MARKUP_KEY, it) }
        disableNotification?.let { putBool(ApiDict.DISABLE_NOTIFICATION_KEY, it) }
        replyToMessageId?.let { putLong(ApiDict.REPLY_TO_MESSAGE_ID_KEY, it) }
        putLong(ApiDict.CHAT_ID_KEY, chatId)
        parseMode?.let { putString(ApiDict.PARSE_MODE_KEY, it) }
        caption?.let { putString(ApiDict.CAPTION_KEY, it) }
    }

    @JsonProperty(ApiDict.METHOD_KEY)
    override val method = "sendAnimation"

}
